x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/repositories/PutCcrRestoreSessionAction.java

```diff
@@ -0,0 +1,126 @@
+
+import java.io.IOException;
+
+public class PutCcrRestoreSessionAction extends Action<PutCcrRestoreSessionAction.PutCcrRestoreSessionResponse> {
+
+    public static final PutCcrRestoreSessionAction INSTANCE = new PutCcrRestoreSessionAction();
+    private static final String NAME = "internal:admin/ccr/restore/session/put";
+
+    private PutCcrRestoreSessionAction() {
+        super(NAME);
+    }
+
+    @Override
+    public PutCcrRestoreSessionResponse newResponse() {
+        return new PutCcrRestoreSessionResponse();
+    }
+
+    @Override
+    public Writeable.Reader<PutCcrRestoreSessionAction.PutCcrRestoreSessionResponse> getResponseReader() {
+        return PutCcrRestoreSessionAction.PutCcrRestoreSessionResponse::new;
+    }
+
+    public static class TransportPutCcrRestoreSessionAction
+        extends TransportSingleShardAction<PutCcrRestoreSessionRequest, PutCcrRestoreSessionResponse> {
+
+        private final IndicesService indicesService;
+        private final CcrRestoreSourceService ccrRestoreService;
+
+        @Inject
+        public TransportPutCcrRestoreSessionAction(ThreadPool threadPool, ClusterService clusterService, ActionFilters actionFilters,
+                                                   IndexNameExpressionResolver resolver, TransportService transportService,
+                                                   IndicesService indicesService, CcrRestoreSourceService ccrRestoreService) {
+            super(NAME, threadPool, clusterService, transportService, actionFilters, resolver, PutCcrRestoreSessionRequest::new,
+                ThreadPool.Names.GENERIC);
+            this.indicesService = indicesService;
+            this.ccrRestoreService = ccrRestoreService;
+        }
+
+        @Override
+        protected PutCcrRestoreSessionResponse shardOperation(PutCcrRestoreSessionRequest request, ShardId shardId) throws IOException {
+            IndexShard indexShard = indicesService.getShardOrNull(shardId);
+            if (indexShard == null) {
+                throw new ShardNotFoundException(shardId);
+            }
+            ccrRestoreService.openSession(request.getSessionUUID(), indexShard);
+            return new PutCcrRestoreSessionResponse(indexShard.routingEntry().currentNodeId());
+        }
+
+        @Override
+        protected PutCcrRestoreSessionResponse newResponse() {
+            return new PutCcrRestoreSessionResponse();
+        }
+
+        @Override
+        protected boolean resolveIndex(PutCcrRestoreSessionRequest request) {
+            return false;
+        }
+
+        @Override
+        protected ShardsIterator shards(ClusterState state, InternalRequest request) {
+            final ShardId shardId = request.request().getShardId();
+            return state.routingTable().shardRoutingTable(shardId).primaryShardIt();
+        }
+    }
+
+
+    public static class PutCcrRestoreSessionResponse extends ActionResponse {
+
+        private String nodeId;
```

Comment:
can this be made final? I see that you both implemented a constructor with StreamInput and the readFrom method?

Reply:
I don't think so. Unfortunately you must implement this:

```
        @Override
        protected PutCcrRestoreSessionResponse newResponse() {
            return new PutCcrRestoreSessionResponse();
        }
```

on `TransportSingleShardAction`.